guidance/chat.py

```diff
@@ -166,7 +166,7 @@ def get_role_end(self, role_name=None):
 # @@@@ Phi-3 @@@@
 # --------------------------------------------------
 # [05/08/24] https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/blob/main/tokenizer_config.json#L119
-phi3_template = "{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') %}{{'<|user|>' + '\n' + message['content'] + '<|end|>' + '\n' + '<|assistant|>' + '\n'}}{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '\n'}}{% endif %}{% endfor %}"
+phi3_template = "{% for message in messages %}{% if message['role'] == 'system' %}{{'<|system|>\n' + message['content'] + '<|end|>\n'}}{% elif message['role'] == 'user' %}{{'<|user|>\n' + message['content'] + '<|end|>\n'}}{% elif message['role'] == 'assistant' %}{{'<|assistant|>\n' + message['content'] + '<|end|>\n'}}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ '<|assistant|>\n' }}{% else %}{{ eos_token }}{% endif %}"
```

Comment:
It looks like the phi-3 mini/small/medium tokenizers have now deviated from each other, where small/medium seem to be following the "old" chat format. Instead of updating the template in-place like this, I would create a new cache entry for the updated mini entry so that we still have compatibility with the small/medium models

Reply:
Actually, funny enough, the medium and small also have slight differences with the presence of the {{ bos_token }} at the start (which makes no practical difference in how guidance uses the template), so we should also have the small/medium templates point to the same chat classes as an alias

Reply:
Just spotted that myself, and have already updated.